services/src/main/java/org/keycloak/services/resources/admin/GroupsResource.java

```diff
@@ -76,18 +77,29 @@ public GroupsResource(RealmModel realm, KeycloakSession session, AdminPermission
                                                @QueryParam("first") Integer firstResult,
                                                @QueryParam("max") Integer maxResults) {
         auth.groups().requireList();
-
         List<GroupRepresentation> results;
 
-        if (Objects.nonNull(search)) {
-            results = ModelToRepresentation.searchForGroupByName(realm, search.trim(), firstResult, maxResults);
-        } else if(Objects.nonNull(firstResult) && Objects.nonNull(maxResults)) {
-            results = ModelToRepresentation.toGroupHierarchy(realm, false, firstResult, maxResults);
+        if (auth.users().canView()) {
+            if (Objects.nonNull(search)) {
+                results = ModelToRepresentation.searchForGroupByName(realm, search.trim(), firstResult, maxResults);
+            } else if(Objects.nonNull(firstResult) && Objects.nonNull(maxResults)) {
+                results = ModelToRepresentation.toGroupHierarchy(realm, false, firstResult, maxResults);
+            } else {
+                results = ModelToRepresentation.toGroupHierarchy(realm, false);
+            }
         } else {
-            results = ModelToRepresentation.toGroupHierarchy(realm, false);
-        }
+            Set<String> groups = auth.groups().getGroupsWithViewPermission();
+            Stream<String> s = groups.stream();
+            if (search!=null) s=s.filter(group->group.contains(search.trim()));
```

Comment:
Don't you have to search by name? Also, I think you should consider filtering by name just like the `[code](https://github.com/pedroigor/keycloak/commit/81e718bf9af893d335c4cb50ddc369c813b0ed26)` I shared with you.